### PR TITLE
Created a new branch as it changes some existing stuff -- please revi…

### PR DESCRIPTION
**Pull request by hftata**
…ew before merging with main branch (I don't trust myself to do it)

### CODE REVIEW

File: R/Weighted_logistic_regression.R [hunk 1]
```diff
@@ -89,20 +89,13 @@ AMS_hb = function(truth, response) {
   s  <- conf.mat[2,2]/(conf.mat[2,2]+conf.mat[1,2]) #TPR - sensitivity
   b  <- conf.mat[2,1]/(conf.mat[2,1]+conf.mat[1,1])
```
Comment (hsansford1) [inline R/Weighted_logistic_regression.R:90]: Think we can delete `AMS_hb` as its the wrong ams

File: R/useful_functions.R [hunk 2]
```diff
@@ -80,22 +137,40 @@ threshold_CV <- function(df, label, weights, theta_0, theta_1, k, n=50){
...
+# takes truth (actual outcome), response (predicted outcome), and weight vector
+# returns AMS
+
+# truth & response can be character ('s' & 'b'),
+#                         factor    (levels named 's' & 'b')
+#                         logical   (TRUE = 's', FALSE = 'b')
+#                         numeric   (1 = 's',    0 = 'b')
+# both vectors should be same type
+
+  if(is.character(truth) | is.factor(truth)){
```
Comment (hsansford1) [inline R/useful_functions.R:164]: Pretty sure at the top of `weighted_logistic_regression.R` we have the label as a factor variable of 1's and 0's, which would make the function fail here?

File: R/Weighted_logistic_regression.R [hunk 3]
```diff
@@ -178,13 +172,14 @@ N_b <- sum(weights[df_train$Label == 0])
...
 weights_Valid <- weights[-trainIndex]
 
-# function for reweighting
-reweight <- function(weights, labels, N_s, N_b){
-  new_weights <- weights
-  new_weights[labels == 1] <- weights[labels == 1] * N_s / sum(weights[labels == 1])
-  new_weights[labels == 0] <- weights[labels == 0] * N_b / sum(weights[labels == 0])
-  return(new_weights)
-}
+# # Delete?? Already in useful_functions
```
Comment (hsansford1) [inline R/Weighted_logistic_regression.R:175]: Sounds good, make sure it is imported here?

File: R/train_test_split.R [hunk 4]
```diff
@@ -21,8 +21,25 @@ library(tidyverse)
...
+
+# To 11 s.f., the sums of weights of subsets
+# KaggleSet=='t', KaggleSet=='b', KaggleSet=='v' all agree with these values
+
+# I'll hard-code Ns, Nb as functions in useful_functions.R, so they can be
+# used when reweighting, etc.
+
 training <- filter(data, KaggleSet == 't')
-training <- select(training, -c('KaggleSet', 'KaggleWeight'))
+training <- select(test, -c('KaggleSet', 'KaggleWeight'))
```
Comment (hsansford1) [inline R/train_test_split.R:42]: this should be training

File: R/useful_functions.R [hunk 5]
```diff
@@ -1,15 +1,67 @@
...
+  # if using hardcoded Ns, Nb makes AMS values work,
+  # remove N_s, N_b function arguments later
+
   new_weights <- weights
   new_weights[labels == 1] <- weights[labels == 1] * N_s / sum(weights[labels == 1])
   new_weights[labels == 0] <- weights[labels == 0] * N_b / sum(weights[labels == 0])
   return(new_weights)
 }
 
+# 'reweight' seems not to do anything...
```
Comment (hsansford1) [inline R/useful_functions.R:21]: reweight works for me, perhaps you are using it on already reweighted data? you can check it has worked by checking `sum(weights[labels == 1]` equals `N_b`

File: R/useful_functions.R [hunk 6]
```diff
@@ -1,15 +1,67 @@
...
+# e.g.:
+load('./data/training.RData')
+reweighted_training <- training
+reweighted_training$reweight1 <- reweight(training$Weight, training$Label, Ns(), Nb())
+reweighted_training$reweight1 == reweighted_training$Weight
+sum(filter(reweighted_training, Label=='s')$reweight1) # != Ns
+# new_weights no different
+
+reweight2 <- function(weights, labels){
```
Comment (hsansford1) [inline R/useful_functions.R:30]: I'm pretty sure `reweight` works fine

File: R/useful_functions.R (hunk 5 above)
Comment (hsansford1) [reply R/useful_functions.R:21]: I see what has gone wrong - you are inputting labels with `'s'` and `'b'` whereas everywhere else we have been using 1 and 0, as is in the function. I think we should just still with one everywhere?

File: R/useful_functions.R (hunk 6 above)
Comment (hsansford1) [inline R/useful_functions.R:30]: I don't think we need this function (if you want, can add cases into the previous `reweight` function for when labels are strings rather than numeric/boolean (note factor can be either 0/1 or s/b)

**Review by hsansford1 — CHANGES_REQUESTED** [review]
I've added some comments, hope you can see them!
